x/auth/vesting/msg_server.go

```diff
@@ -99,10 +99,183 @@ func (s msgServer) CreateVestingAccount(goCtx context.Context, msg *types.MsgCre
+		lockupCoins = lockupCoins.Add(period.Amount...)
+	}
+
+	// if lockup absent, default to an instant unlock schedule
+	lockupPeriods := msg.LockupPeriods
+	vestingPeriods := msg.VestingPeriods
+	if !vestingCoins.IsZero() && len(msg.LockupPeriods) == 0 {
+		lockupPeriods = []types.Period{
+			{Length: 0, Amount: vestingCoins},
+		}
+		lockupCoins = vestingCoins
+	}
+
+	if !lockupCoins.IsZero() && len(msg.VestingPeriods) == 0 {
+		// If vesting absent, default to an instant vesting schedule
+		vestingPeriods = []types.Period{
+			{Length: 0, Amount: lockupCoins},
+		}
+		vestingCoins = lockupCoins
+	}
+
+	// The vesting and lockup schedules must describe the same total amount.
+	// IsEqual can panic, so use (a == b) <=> (a <= b && b <= a).
+	if !types.CoinEq(lockupCoins, vestingCoins) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "lockup and vesting amounts must be equal")
+	}
+
+	var madeNewAcc bool
+	acc := ak.GetAccount(ctx, to)
+	var vestingAcc *types.ClawbackVestingAccount
+
+	// a grant can be added only if toAddress exists, msg.Merge && isClawback && to.FunderAddress == msg.FromAddress
+	if acc != nil {
+		var isClawback bool
+		vestingAcc, isClawback = acc.(*types.ClawbackVestingAccount)
+		switch {
+		case !isClawback:
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrNotSupported, "account %s must be a clawback vesting account", msg.ToAddress)
+		case !msg.Merge && isClawback:
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "account %s already exists; consider setting  'merge' to 'true'", msg.ToAddress)
+		case msg.FromAddress != vestingAcc.FunderAddress:
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "account %s can only accept grants from account %s", msg.ToAddress, vestingAcc.FunderAddress)
+		}
+		grantAction := types.NewClawbackGrantAction(msg.FromAddress, msg.StartTime, msg.GetLockupPeriods(), msg.GetVestingPeriods(), vestingCoins)
+		err := vestingAcc.AddGrant(ctx, grantAction)
+		if err != nil {
+			return nil, err
+		}
+		ak.SetAccount(ctx, vestingAcc)
+	} else {
+		baseAcc := authtypes.NewBaseAccountWithAddress(to)
+		vestingAcc = types.NewClawbackVestingAccount(
+			baseAcc,
+			from,
+			vestingCoins,
+			msg.StartTime,
+			lockupPeriods,
+			vestingPeriods,
+		)
+		acc := ak.NewAccount(ctx, vestingAcc)
+		madeNewAcc = true
+		ak.SetAccount(ctx, acc)
+	}
+
+	if madeNewAcc {
+		defer func() {
+			telemetry.IncrCounter(1, "new", "account")
+
+			for _, a := range vestingCoins {
+				if a.Amount.IsInt64() {
+					telemetry.SetGaugeWithLabels(
+						[]string{"tx", "msg", "create_clawback_vesting_account"},
+						float32(a.Amount.Int64()),
+						[]metrics.Label{telemetry.NewLabel("denom", a.Denom)},
+					)
+				}
+			}
+		}()
+	}
+
+	err = bk.SendCoins(ctx, from, to, vestingCoins)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
```

Comment:
I think it would be helpful if we emitted additional events about this. E.g. the address and amount should suffice. WDYT?

Reply:
yeah totally agreed!